### PR TITLE
machine-api: disallow unsetting authoritativeAPI

### PR DESCRIPTION
**Pull request by damdo**
We should disallow unsetting authoritativeAPI once it has been set.

### CODE REVIEW

File: machine/v1beta1/types_machineset.go [hunk 1]
```diff
@@ -112,6 +112,7 @@ type MachineTemplateSpec struct {
 
 // MachineSetStatus defines the observed state of MachineSet
 // +openshift:validation:FeatureGateAwareXValidation:featureGate=MachineAPIMigration,rule="!has(oldSelf.synchronizedGeneration) || (has(self.synchronizedGeneration) && self.synchronizedGeneration >= oldSelf.synchronizedGeneration) || (oldSelf.authoritativeAPI == 'Migrating' && self.authoritativeAPI != 'Migrating')",message="synchronizedGeneration must not decrease unless authoritativeAPI is transitioning from Migrating to another value"
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=MachineAPIMigration,rule="self.authoritativeAPI != ” || oldSelf.authoritativeAPI == ”",message="authoritativeAPI cannot be unset once it has been set"
```
Comment (JoelSpeed) [inline machine/v1beta1/types_machineset.go:115]: Do we also want something similar on the spec? What about for Machines?

I would like to see some testing with the integration test suite that tests this validation transition

**Comment by JoelSpeed** [comment]
@damdo Should we pick this back up?

**Comment by damdo** [comment]
Yes

**Comment by JoelSpeed** [comment]
@damdo Are we ready to move this one forward?

File: machine/v1beta1/types_machine.go [hunk 2]
```diff
@@ -220,6 +220,7 @@ type Machine struct {
 }
 
 // MachineSpec defines the desired state of Machine
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=MachineAPIMigration,rule="self.?authoritativeAPI.orValue('') != '' || oldSelf.?authoritativeAPI.orValue('') == ''",message="authoritativeAPI cannot be unset once it has been set"
```
Comment (JoelSpeed) [inline machine/v1beta1/types_machine.go]: The spec field is defaulted, so, this won't ever trigger will it? Can you demonstrate through the integration test that this triggers?

Comment (JoelSpeed) [inline machine/v1beta1/types_machine.go]: Why do we need the optional old self here? Does the following not work? And would be simpler?

```
has(self.authoritativeAPI) || !has(oldSelf.authoritativeAPI)
```

Comment (damdo) [reply machine/v1beta1/types_machine.go]: The `has()` approach would allow an explicit empty string while `!= ''` would block it.
But that likely doesn't matter here because the enum validation `Enum=MachineAPI;ClusterAPI` already rejects empty strings. So both approaches achieve the same end result

`self.?authoritativeAPI.orValue('') != '' || oldSelf.?authoritativeAPI.orValue('') == ''`
Checks if the value is non-empty and so it needs`?.orValue('')` to handle missing fields without errors

`has(self.authoritativeAPI) || !has(oldSelf.authoritativeAPI)`
Checks if the field key exists in the object and `has()` handles missing fields (returning false, no error)

So I guess both work, but maybe the latter is cleaner?

Comment (damdo) [reply machine/v1beta1/types_machine.go]: Are you referring to specific API integration tests in this repo or on CCAPIO?

Comment (JoelSpeed) [reply machine/v1beta1/types_machine.go]: The tests in this repo

Comment (JoelSpeed) [reply machine/v1beta1/types_machine.go]: The former confuses two responsibilities. The latter has a single responsibility.

The emptiness check is handled at the field level by the enum. The empty string isn't a valid value so we don't need to check it in this CEL

Comment (JoelSpeed) [reply machine/v1beta1/types_machine.go]: The former confuses two responsibilities. The latter has a single responsibility.

The emptiness check is handled at the field level by the enum. The empty string isn't a valid value so we don't need to check it in this CEL

Comment (damdo) [reply machine/v1beta1/types_machine.go]: Ok happy to give the second one a try, as it works on principle and looks simpler.

Comment (damdo) [reply machine/v1beta1/types_machine.go]: Updated with changed approach.

Comment (damdo) [reply machine/v1beta1/types_machine.go]: 👍 tests added. Let's see what they report

File: machine/v1beta1/tests/machines.machine.openshift.io/MachineAPIMigration.yaml [hunk 3]
```diff
@@ -77,6 +77,17 @@ tests:
...
+    initial: |
+      apiVersion: machine.openshift.io/v1beta1
+      kind: Machine
+      spec:
+        authoritativeAPI: MachineAPI
+    updated: |
+      apiVersion: machine.openshift.io/v1beta1
+      kind: Machine
+      spec: {}
+    expectedError: "spec: Invalid value: \"object\": authoritativeAPI cannot be unset once it has been set"
```
Comment (JoelSpeed) [inline machine/v1beta1/tests/machines.machine.openshift.io/MachineAPIMigration.yaml]: Yeah so this test failed because of the defaulting, we don't need the rule on the spec, only status

File: machine/v1beta1/types_machine.go (hunk 2 above)
Comment (damdo) [reply machine/v1beta1/types_machine.go]: It looks like the [new tests are failing](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_api/2091/pull-ci-openshift-api-master-integration/2011092993234178048)

I think it is because the `spec.authoritativeAPI` field has a default value (MachineAPI).
When the test tries to remove the field, the defaulting mechanism automatically re-adds it with the default value.
So the validation rule never triggers because the field is never actually absent.

Should the test instead try to set it as empty string `''`?

Comment (JoelSpeed) [reply machine/v1beta1/types_machine.go]: I don't think we need to test the string being set to empty, I trust the enum validation that is in place. I think the spec side of this change can be removed and we can just focus on the status side

File: machine/v1beta1/types_machine.go [hunk 4]
```diff
@@ -317,6 +317,7 @@ type LifecycleHook struct {
 
 // MachineStatus defines the observed state of Machine
 // +openshift:validation:FeatureGateAwareXValidation:featureGate=MachineAPIMigration,rule="!has(oldSelf.synchronizedGeneration) || (has(self.synchronizedGeneration) && self.synchronizedGeneration >= oldSelf.synchronizedGeneration) || (oldSelf.authoritativeAPI == 'Migrating' && self.authoritativeAPI != 'Migrating')",message="synchronizedGeneration must not decrease unless authoritativeAPI is transitioning from Migrating to another value"
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=MachineAPIMigration,rule="has(self.authoritativeAPI) || !has(oldSelf.authoritativeAPI)",message="authoritativeAPI cannot be unset once it has been set"
```
Comment (JoelSpeed) [inline machine/v1beta1/types_machine.go]: Nit: Can we change the wording here to be consistent with other places we have similar?


```suggestion
// +openshift:validation:FeatureGateAwareXValidation:featureGate=MachineAPIMigration,rule="has(self.authoritativeAPI) || !has(oldSelf.authoritativeAPI)",message="authoritativeAPI may not be removed once set9+6*-"
```

Comment (damdo) [reply machine/v1beta1/types_machine.go]: Done

**Comment by JoelSpeed** [comment]
/lgtm
/label acknowledge-critical-fixes-only
/verified by integration tests

This PR doesn't actually effect the payload so no need to block on ack
